stdlib/list.mli

```diff
@@ -141,6 +141,12 @@ val filter_map : ('a -> 'b option) -> 'a list -> 'b list
     @since 4.08.0
 *)
 
+val flat_map : ('a -> 'b list) -> 'a list -> 'b list
+(** [flat_map f l] is equivalent to [flatten (map f l)].
```

Comment:
I think it's worth mentioning that this function is tail-recursive, which is at odds with both `flatten` and `map`.  The docstring for `rev_map` sets a precedent: 
````ocaml
(** [List.rev_map f l] gives the same result as
   {!List.rev}[ (]{!List.map}[ f l)], but is tail-recursive and
   more efficient. *)
````
(except that here it would probably be "but is tail-recursive and less efficient")

Reply:
Unless it's a strong requirement on your side, I would rather not. I think it makes sense to mention it for `rev_map` because the natural implementation is simple and tail-recursive -- in a sense it is part of the specification of lists that rev_map is easier than map. Here being tail-recursive is not a natural choice (the simplest implementation isn't -- until TRMC is added to the system, and then it may still not be) but rather an implementation choice that I made to be allowed to not bother with this. It is more similar to `List.init` (anecdotally tail-recursive), for which the current choice also isn't documented.

My general plan for this is (1) get TRMC and (2) make sure every standard-library List function is tail-recursive and be done with the whole thing. In the meanwhile, I'll do lip-service to having code that actually works for users if it doesn't make the implementation too ugly, and experts are always free to check the implementation and use an alternative library if they don't like it.

Reply:
If you don't get to (1) for 4.10 I would prefer to know that the function won't fail on big inputs (but you can leave the "less efficient" bit out as far as I'm concerned).

Reply:
>It is more similar to List.init (anecdotally tail-recursive), for which the current choice also isn't documented.

There is no "anecdotally" here. Every function in `List` is (supposed to be) documented as tail-recursive or not. The default is tail-rec, as documented by the first paragraph of `List.mli` and every non-tail-rec function is documented as such.

IOW it's not an implementation choice, it's part of the interface.

Reply:
So if I understand correctly, I was correct not to write anything in the function specification given that it respects the default of being tail-recursive. I'll mark this conversation resolved.

Reply:
I don't want to be overly pedantic, but when I read that `[flat_map f l] is equivalent to [flatten (map f l)]` and I know that both `map` and `flatten` can blow the stack, I deduce that `[flat_map f l]` can blow the stack as well.  What about

`[flat_map f l] is equivalent to [flatten (map f l)], except that it is tail-recursive`